### ISSUE TITLE
[FSDP] summon_full_params gradient access issues

### ISSUE THREAD

**Issue by rohan-varma** [pytorch/pytorch#73891]
### 🐛 Describe the bug

Under a summon_full_params context, there are issues when accessing .grad field of parameters from parameters() or named_parameters() function.

For example, in the following test:
```
def test_params_are_unflattenned(self):
        layer_shape = (10, 12)
        model = nn.Linear(*layer_shape, bias=False).cuda(self.rank)
        fsdp_model = FSDP(deepcopy(model)).cuda(self.rank)

        flattened_param = fsdp_model.get_parameter("_fsdp_wrapped_module.flat_param")
        self.assertEqual(layer_shape[0] * layer_shape[1] / 2, flattened_param.numel())


        input = torch.randn(1, 10, device=self.rank)

        out = fsdp_model(input).sum()
        out.backward()
        with fsdp_model.summon_full_params():
            named_parameters = fsdp_model.named_parameters()
            if self.rank == 0:
                np = dict(named_parameters)
                for name, param in np.items():
                    print(f"name: {name} param grad {param.grad}")

                params = list(fsdp_model.parameters())
                for p in params:
                    print(f"param grad: {p.grad}")

            self.assertEqual(fsdp_model.weight.shape, model.weight.shape)
```

for named_parameters() this results in:

```
name: _fsdp_wrapped_module._fpw_module.weight param grad None
```

and for `parameters()`, this results in just `None` for param.grad.

For both function, `.grad` attribute works well outside of the context, i.e. user gets back the sharded gradient.

### Versions

main

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

**Comment by awgu** [comment]
An initial PR: https://github.com/pytorch/pytorch/pull/85738
The API may be rounded out in future follow-ups.
